torch/fx/graph_module.py

```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
         cls = type(self)
         cls.forward = _forward_from_src(self._code)
 
-        cls_call = cls.__call__
+        # If we're calling `recompile` on a GraphModule that has already
+        # been recompiled, `cls.__call__` would be `wrapped_call`.
+        # We don't want to wrap `wrapped_call` again (as this would
+        # ultimately lead to O(num_recompiles) "unwrappings" later), so
+        # we use the instance attribute `_cls_call` to store the
+        # original implementation of `__call__`
+        if not self._cls_call:
+            self._cls_call = cls.__call__
 
         def print_full_traceback(exctype, value, tb):
             traceback.print_exception(exctype, value, tb)
 
+        # Previously, if an error occurred when valid
+        # symbolically-traced code was run with an invalid input, the
+        # user would see the source of the error as coming from
+        # `File "<eval_with_key_N">`, where N is the number of times
+        # `recompile` had been called on the GraphModule. We use this
+        # function to generate a more informative error message. We
+        # return the traceback itself, a message explaining that the
+        # error occurred in a traced Module's generated forward
+        # function, and five lines of context surrounding the faulty
+        # faulty line
+        def generate_error_message(tb: TracebackType) -> str:
+            # auxiliary variables (for readability)
+            frame_summary = traceback.extract_tb(tb)[-1]
+            err_lineno = frame_summary.lineno
+            err_line_len = len(frame_summary.line)
+            all_src_lines = list(_eval_cache.values())[-1]
```

Comment:
Pulling off the back of the eval cache is not the right thing to do here. In python, global variable references are looked up on every call and not closed over. This creates an incorrect context:

```
import torch
import torch.fx

class M(torch.nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

traced_dot = torch.fx.symbolic_trace(M())

class N(torch.nn.Module):
    def forward(self, x):
        return torch.relu(x)

traced_relu = torch.fx.symbolic_trace(N())

traced_dot(5)
"""
Traceback (most recent call last):
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 350, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/data/users/jamesreed/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 4, in forward
    dot_1 = torch.dot(w, x);  w = x = None
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
Call to “exec” using an FX-traced Module, line 4 of the traced Module’s generated forward function:
    relu_1 = torch.relu(x);  x = None
    return relu_1
    
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
Traceback (most recent call last):
  File "eval_cache.py", line 20, in <module>
    traced_dot(5)
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 358, in wrapped_call
    raise e.with_traceback(None)
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
"""

I think a simple thing to do here instead is `all_src_lines = _eval_cache[frame_summary.filename]`
```